### PR TITLE
0.96.7 BIS Support modules for artillery

### PR DESCRIPTION
**Pull request by Wyqer**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| Needs wipe? | no |

### Description:
Implementation of the [Support modules framework from Arma 2/Arma3](https://community.bistudio.com/wiki/Simple_Support_Module) for Liberation.
Worked fine so far in local hosted. Dedicated test will be done, after this and #754 is merged.
It seems, I was already very close three years ago with this try: https://github.com/KillahPotatoes/KP-Liberation/commit/d090744af19b52791b066336485c168adb69e668

### Content:
- [x] Artillery support module functionality
- [x] Mission parameter setting
- [x] Whitelist array for players and valid units

### Successfully tested on:
- [x] Local MP
- [x] Dedicated MP

### CODE REVIEW

**Comment by Wyqer** [comment]
Worked fine so far in dedicated server environment.
Some smaller things like one didn't get the radio commands while other do (after some mission time). All of these aren't critical and the functionality itself worked fine for everyone. Guess we need to tweak it later after some more user tests/reports, but for a first shipment it should be good.
In local hosted it's working totally fine. At least in my tests so far.

File: Missionframework/onPlayerRespawn.sqf [hunk 1]
```diff
@@ -16,3 +18,28 @@ if (isNil "GRLIB_respawn_loadout") then {
...
+    player isEqualTo ([] call KPLIB_fnc_getCommander),
+    (getPlayerUID player) in KP_liberation_suppMod_whitelist,
+    true
+] select KP_liberation_suppMod) then {
+    waitUntil {!isNil "KPLIB_suppMod_req" && !isNil "KPLIB_suppMod_arty" && time > 5};
+
+    // Remove link to corpse, if respawned
+    if (!isNull _oldUnit) then {
+        KPLIB_suppMod_req synchronizeObjectsRemove [_oldUnit];
+        _oldUnit synchronizeObjectsRemove [KPLIB_suppMod_req];
```
Comment (veteran29) [inline Missionframework/onPlayerRespawn.sqf:34]: Isn't the synchronization two way? I don't think you need to link/unlink both sides.

File: Missionframework/scripts/server/highcommand/fn_highcommand.fsm [hunk 2]
```diff
@@ -294,7 +295,11 @@ class FSM
                         itemno = 12;
                         init = /*%FSM<STATEINIT""">*/"{" \n
                          "    _cmdr hcRemoveGroup _x;" \n
-                         "} forEach (_hcGrps select {((units _x) select {alive _x}) isEqualTo []});"/*%FSM</STATEINIT""">*/;
+                         "} forEach (_hcGrps select {" \n
+                         "    ((units _x) select {alive _x}) isEqualTo []" \n
```
Comment (veteran29) [inline Missionframework/scripts/server/highcommand/fn_highcommand.fsm:299]:
```suggestion
                         "    ((units _x) findIf {alive _x}) == -1" \n
```

File: Missionframework/scripts/server/support/fn_createSuppModules.sqf [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    waitUntil {!isNil "save_is_loaded" && {save_is_loaded}};
+    ["Init provider on server", "SUPPORTMODULES"] call KPLIB_fnc_log;
+    [KPLIB_suppMod_req] call BIS_fnc_moduleSupportsInitRequester;
+    [KPLIB_suppMod_arty] call BIS_fnc_moduleSupportsInitProvider;
+
+    // Hide the three HQ entities created at zero pos. BIS scripts only hides them local for the creator
+    waitUntil {!isNil "BIS_SUPP_HQ_WEST" && !isNil "BIS_SUPP_HQ_EAST" && !isNil "BIS_SUPP_HQ_GUER"};
+    {
+        hideObjectGlobal _x;
+    } forEach [BIS_SUPP_HQ_WEST, BIS_SUPP_HQ_EAST, BIS_SUPP_HQ_GUER]
```
Comment (veteran29) [inline Missionframework/scripts/server/support/fn_createSuppModules.sqf:47]: Where are these created? Is this done by some bis fnc?

File: Missionframework/scripts/shared/fetch_params.sqf [hunk 4]
```diff
@@ -429,6 +430,15 @@ if (!isDedicated && hasInterface) then {
     _value = if (KP_liberation_high_command) then {localize "STR_PARAMS_ENABLED";} else {localize "STR_PARAMS_DISABLED";};
     _text = _text + format ["<font color='#ff8000'>%1</font><br />%2<br /><br />", _param, _value];
 
+    _param = localize "STR_PARAM_SUPPMOD";
+    switch (KP_liberation_victoryCondition) do {
```
Comment (veteran29) [inline Missionframework/scripts/shared/fetch_params.sqf:434]:
```suggestion
    switch (KP_liberation_suppMod) do {
```

File: Missionframework/stringtable.xml [hunk 5]
```diff
@@ -6955,6 +6955,22 @@
...
+            <Original>BI Support System access</Original>
+            <German>Zugriff auf BI Support System</German>
+        </Key>
+        <Key ID="STR_PARAM_SUPPMOD_COMMANDER">
+            <Original>Commander only</Original>
+            <German>Nur Kommandant</German>
+        </Key>
+        <Key ID="STR_PARAM_SUPPMOD_CMDRANDWHITELIST">
+            <Original>Commander and Whitelist only</Original>
+            <German>Nur Kommandant und Whitelist</German>
```
Comment (veteran29) [inline Missionframework/stringtable.xml:6968]: I'm not sure if we need `command` and `commander and whitelist` params. Whitelist is empty by default so it's kinda the same as `commander` only by default.

File: Missionframework/onPlayerRespawn.sqf (hunk 1 above)
Comment (Wyqer) [reply Missionframework/onPlayerRespawn.sqf:34]: Yeah, normally it is. I've also just done this one way for the HC implementation.
I used this in both directions at this place, as the `BIS_fnc_addSupportLink` does it also internally. So a "shouldn't be necessary, but doesn't hurt and as BIS use it in their function, better safe then sorry"

File: Missionframework/scripts/server/support/fn_createSuppModules.sqf (hunk 3 above)
Comment (Wyqer) [reply Missionframework/scripts/server/support/fn_createSuppModules.sqf:47]: Yeah, they are created in the BIS init functions for the support modules.
They are the 3 single soldiers which could be seen at 0,0,0 pos when using this module.

File: Missionframework/scripts/shared/fetch_params.sqf (hunk 4 above)
Comment (Wyqer) [reply Missionframework/scripts/shared/fetch_params.sqf:434]: Oh ^^

File: Missionframework/stringtable.xml (hunk 5 above)
Comment (Wyqer) [reply Missionframework/stringtable.xml:6968]: So removing the "Only Commander"?
Should be fine.

File: Missionframework/onPlayerRespawn.sqf (hunk 1 above)
Comment (veteran29) [reply Missionframework/onPlayerRespawn.sqf:34]: 👍

File: Missionframework/scripts/client/build/do_build.sqf [hunk 6]
```diff
@@ -320,21 +320,25 @@ while { true } do {
                 } else {
                     _vehicle setVectorUp surfaceNormal position _vehicle;
                 };
+                
+                // Add artillery to provider module, if support module is enabled
+                if (KP_liberation_suppMod > 0 && {_classname in KP_liberation_suppMod_artyVeh}) then {
```
Comment (veteran29) [inline Missionframework/scripts/client/build/do_build.sqf:325]: We could have this in object inits.

File: Missionframework/scripts/server/game/save_manager.sqf [hunk 7]
```diff
@@ -341,9 +343,9 @@ if (!isNil "_saveData") then {
             _object setPosWorld _pos;
             _object setVectorDirAndUp [_vecDir, _vecUp];
 
-            // Add blufor crew, if it had crew or is a UAV
-            if ((unitIsUAV _object) || _hascrew) then {
-                [_object] call KPLIB_fnc_forceBluforCrew;
+            // Add artillery to provider module, if support module is enabled
+            if (KP_liberation_suppMod > 0 && {_class in KP_liberation_suppMod_artyVeh}) then {
```
Comment (veteran29) [inline Missionframework/scripts/server/game/save_manager.sqf:347]: We could have this in object inits.

File: Missionframework/scripts/server/highcommand/fn_highcommand.fsm [hunk 8]
```diff
@@ -294,7 +295,11 @@ class FSM
                         itemno = 12;
                         init = /*%FSM<STATEINIT""">*/"{" \n
                          "    _cmdr hcRemoveGroup _x;" \n
-                         "} forEach (_hcGrps select {((units _x) select {alive _x}) isEqualTo []});"/*%FSM</STATEINIT""">*/;
+                         "} forEach (_hcGrps select {" \n
+                         "    ((units _x) findIf {alive _x}) == -1" \n
+                         "    || _cmdr in (units _x)" \n
+                         "    || (missionNamespace getVariable [""BIS_SUPP_HQ_WEST"", nil]) in (units _x)" \n
```
Comment (veteran29) [inline Missionframework/scripts/server/highcommand/fn_highcommand.fsm:301]: nil in <ARRAY> will result in nil and break whole expression.
```suggestion
                         "    || (missionNamespace getVariable [""BIS_SUPP_HQ_WEST"", objNull]) in (units _x)" \n
```

File: Missionframework/scripts/server/highcommand/fn_highcommand.fsm [hunk 9]
```diff
@@ -258,6 +258,7 @@ class FSM
                          "    && {!(_x in _hcGrps)}" \n
                          "    && {!(((units _x) select {alive _x}) isEqualTo [])}" \n
                          "    && {!(_cmdr in (units _x))}" \n
+                         "    && {!((missionNamespace getVariable [""BIS_SUPP_HQ_WEST"", nil]) in (units _x))}" \n
```
Comment (veteran29) [inline Missionframework/scripts/server/highcommand/fn_highcommand.fsm:261]:
```suggestion
                         "    && {!((missionNamespace getVariable [""BIS_SUPP_HQ_WEST"", objNull]) in (units _x))}" \n
```

Comment (veteran29) [reply Missionframework/scripts/server/highcommand/fn_highcommand.fsm:261]: Also what if the player is not WEST but EAST?

File: Missionframework/scripts/client/build/do_build.sqf (hunk 6 above)
Comment (Wyqer) [reply Missionframework/scripts/client/build/do_build.sqf]: Done

File: Missionframework/scripts/server/game/save_manager.sqf (hunk 7 above)
Comment (Wyqer) [reply Missionframework/scripts/server/game/save_manager.sqf]: Done

**Pull request merged**